working/0723-static-extensions/feature-specification.md

```diff
@@ -315,165 +453,482 @@ extension.
+reference denotes an invocation of or reference to the constructor
+named `C.name` which is declared by the extension declaration that _M_
+contains.
 
 *Note that no attempt is made to determine that some constructors are "more
 specific" or "less specific", it is simply a conflict if there are multiple
 constructors with the requested name in the accessible extensions.*
 
-### Dynamic Semantics
+Let `E` be the unique extension declaration that _M_ contains.  The
+fully resolved form of the original invocation or reference is then
+`E.name(arguments)` or `E.name`, respectively.
+
+#### Type inference for constructor invocations with no provided type arguments.
+
+Consider an instance creation expression of the form
+`C.name(arguments)`, where `E.name(arguments)` is the fully resolved
+invocation of the constructor.
+
+*In the case that the invocation does not correspond to the invocation
+of a constructor declared in an extension, there is no fully resolved
+invocation and this section does not apply.*
+
+Type inference for such an invocation is done by performing inference
+on the fully resolved invocation `E.name(arguments)` as defined in the
+subsequent section, using the same downwards context as the original
+expression.
+
+*Inference serves to find the type arguments (if any) that are missing
+from the fully resolved invocation.  These arguments are what are
+needed for subsequent static checking and for the dynamic
+semantics. However, for the purposes of error reporting, it may be
+useful to the user to report errors in terms of the original
+syntactic form.  If `C` is a generic type, then the corresponding type
+arguments for `C` can be obtained simply by substituting the inferred
+type arguments to `E` for the type parameters of `E` in the on-type of
+`E` (which by construction is an instantation of `C`).*
+
+The static type of the constructor invocation in this case is the
+fully instantiated on-type of `E` - that is, the on-type of `E` with
+the inferred type arguments of the fully resolved invocation
+substituted for the type parameters of `E`.
+
+#### Type inference for constructor invocations with explicitly provided type arguments.
+
+Consider an instance creation expression of the form
+`C<TypeArguments>.name(arguments)`, where `E.name(arguments)` is the
+fully resolved invocation of the constructor.
+
+Type inference for such an invocation is done by performing inference
+on the fully resolved invocation `E.name(arguments)` as defined in the
+subsequent section, using `C<TypeArguments>` as the downwards context.
+
+*As above, inference serves to find the type arguments that are
+missing from the fully resolved invocation.  For the purposes of error
+reporting, it may be more useful to the user to report errors in terms
+of the invocation form using `C<TypeArguments>` as given in the
+original program.*
+
+The static type of the constructor invocation in this case is
+`C<TypeArguments>`.
```

Comment:
That's a choice. It could also be the instantiated `on` type of `E`, which could be a proper subtype.

Do we choose `C<TypeArguments>` because we expect the user to want that type,
for example as the type of a variable initialized with this value?

(We should check, when we get to tear-offs, that `C<TypeArguments>.name(args)` has the same static type as `(C<TypeArguments>.name)(args)`.)